qemu/tests/blockdev_stream_subchain.py

```diff
@@ -0,0 +1,162 @@
+import logging
+
+from virttest import data_dir
+from virttest import qemu_storage
+
+from provider import backup_utils
+
+from provider.blockdev_stream_base import BlockDevStreamTest
+from provider.virt_storage.storage_admin import sp_admin
+
+
+class BlockdevStreamSubChainTest(BlockDevStreamTest):
+    """Do block-stream based on an existed snapshot in snapshot chain"""
+
+    def __init__(self, test, params, env):
+        super(BlockdevStreamSubChainTest, self).__init__(test, params, env)
+        self._snapshot_images = self.params.objects("snapshot_images")
+        self._base_node_tag = self.params["base_node_tag"]
+        self._trash = []
+
+    def snapshot_test(self):
+        """create one snapshot, create one new file"""
+        self.generate_tempfile(self.disks_info[0][1],
+                               filename="base",
+                               size=self.params["tempfile_size"])
+
+        # data->sn1->sn2->sn3
+        chain = [self.base_tag] + self._snapshot_images
+        for idx in range(1, len(chain)):
+            backup_utils.blockdev_snapshot(
+                self.main_vm,
+                "drive_%s" % chain[idx-1],
+                "drive_%s" % chain[idx]
+            )
+
+            self.generate_tempfile(self.disks_info[0][1],
+                                   filename=chain[idx],
+                                   size=self.params["tempfile_size"])
+
+    def _disk_define_by_params(self, tag):
+        params = self.params.copy()
+        params.setdefault("target_path", data_dir.get_data_dir())
+        return sp_admin.volume_define_by_params(tag, params)
+
+    def prepare_snapshot_file(self):
+        """hotplug all snapshot images"""
+        for tag in self._snapshot_images:
+            disk = self._disk_define_by_params(tag)
+            disk.hotplug(self.main_vm)
+            self._trash.append(disk)
+
+    def _remove_images(self):
+        for img in self._trash:
+            sp_admin.remove_volume(img)
+
+    def post_test(self):
+        try:
+            if self.clone_vm.is_alive():
+                self.clone_vm.destroy()
+            self._remove_images()
+        except Exception as e:
+            logging.warning(str(e))
+
+    def _is_same_file(self, file_params, file_opts):
+        # FIXME: this should be supported in VT
+        if file_opts["driver"] == "gluster":
+            if file_params["path"] != file_opts["path"]:
+                return False
+        elif file_opts["driver"] == "iscsi":
+            if file_params["lun"] != file_opts["lun"]:
+                return False
+        elif file_opts["driver"] == "nbd":
+            if file_params["server.port"] != file_opts["server.port"]:
+                return False
+        elif file_opts["driver"] == "rbd":
+            if file_params["image"] != file_opts["image"]:
+                return False
+        else:
+            if file_params["filename"] != file_opts["filename"]:
+                return False
+        return True
```

Comment:
```suggestion
        mapping = {'gluster': 'path', 'iscsi': 'lun', 
                  'nbd': 'server.port', 'rbd': 'image'}
        for driver, attr in mapping.items():
            if file_opts["driver"] == driver:
                return file_params[attr] == file_opts[attr]
            return file_params["filename"] == file_opts["filename"]
```

Reply:
yes, it's better, I made some code change and will do some tests later, thank you.